### ISSUE TITLE
[Discuss] Supporting hash-based anonymization

### ISSUE THREAD

**Issue by ajosh0504** [elastic/anonymize-it#38]
The current Faker-based anonymization implementation works for one-off anonymization use cases. However, it has the following drawbacks in cases where users might want to correlate anonymized data collected periodically/ using multiple queries/ across multiple sources (clusters):

- Faker mappings are not persistent across multiple runs if the order in which the aggregated results appear while creating the mappings changes between runs. Eg: If we want to correlate login and network activity for a particular host in a given environment and the user makes two separate queries for login and  network events, we can't be sure that the faker mapping for fields like hostname will be maintained across both queries
- Faker mappings might not be persistent across queries run on multiple clusters

An alternative to Faker-based anonymization for such data correlation usecases is to use a hash-based anonymization technique wherein the anonymizer accepts a "password" configured by the user. The user will always use the same password for all future anonymization runs. Fields to be anonymized are replace by the sha256 of a combination of the password and the field value. 

**Discussion items**:

-  What happens if the user forgets the password? Are there other ways in the Elastic framework to get a static but unique representation of a user/environment so as to avoid human error?
- Do we want to support both anonymization techniques or just hash based? How do we go about this?

**Comment by ajosh0504** [comment]
My thoughts on the discussion items:
-  We start with using the password obtained from the user through the config file, while we explore ways around it.
-  If we want to support both techniques, we can make this also configurable, in that case, we have separate Anonymizer classes corresponding to each. Alternatively, we can make changes to the current Anonymizer class to accommodate the hash-based method.

**Comment by ajosh0504** [comment]
To get a better understanding of what I was proposing, [here](https://github.com/elastic/anonymize-it/pull/39) is a PR draft. Feel free to suggest restructuring changes there if that's easier.

**Comment by christophercutajar** [comment]
As discussed, I really like the idea of using hashing to be able to for correlation purposes. This way the anonymization can be execution multiple times while returning the same irreversible hash. Using hash anonymization would also not intefer with ECS as both `host.name` and `hash.sha256` or `hash.sha512` are of type `keyword`.

My thoughts on the discussion items:
- I'm ok with testing with password but I don't think this is the best approach. The drawbacks I see it that,
	- if the clusters are owned by multiple departments, this need to be shared across teams
	- if Elastic requires the customer at a later stage to get more data, there is the possibility that the user didn't store their password
	- the password storage need to be available to all customer internal team due to personnel rotation
- I like to make this configurable same as done in your PR.

Just as note, we would need to re-enagage assurance team to re-review the anonymizer on the new anonymzation algorithm. Ideally we should engage them now to also get their opinion. Since you amazing have POC in place this would not take long. I could start the discussion with the team.

**Comment by ajosh0504** [comment]
@christophercutajar Do we have any alternatives to using a user configured password for the hashing approach at the moment?

**Comment by christophercutajar** [comment]
Something that will not change across all ES clusters and also would not change on a license renewal is the `issued_to` field in the `_license` API. The only time that a customer would have a different `issued_to` is when the customer have a `production` and `non-production` license. Production licenses will have the company name in the `issued_to` field while `non-production` licenses would have the company name together with `(non-production environments)` in the `issued_to` field.

In my view, the anonymizer could make an extra call to the cluster to get the `issued_to` from the `license` and then use this as the salt for the hashing.

**Comment by christophercutajar** [comment]
Checked how licenses are on Elastic Cloud, seems like Elastic Cloud is issued one Elastic license to be used across all customers. In this case the `issued_to` would not work!

```
{
  "license" : {
    "status" : "active",
    "uid" : "e50e4f44-f926-404f-9279-5f5d2fe1fc77",
    "type" : "platinum",
    "issue_date" : "2020-03-24T00:00:00.000Z",
    "issue_date_in_millis" : 1585008000000,
    "expiry_date" : "2021-06-30T00:00:00.000Z",
    "expiry_date_in_millis" : 1625011200000,
    "max_nodes" : 100000,
    "issued_to" : "Elastic Cloud",
    "issuer" : "API",
    "start_date_in_millis" : 1585008000000
  }
}
```

**Comment by ajosh0504** [comment]
If that's the case, we could do the following:

- Add an "account_type" field to the anonymization config, which accepts "on-prem" or "cloud" as the options.
- For on-prem, we use the "issued_to" from the GET /_license API call as the salt
- For cloud, we could either prompt the user to enter the email ID associated with their account or ask support to give the customer their account ID which would be used as the salt. This also means support does not have to maintain a separate set of keys for this particular project.

**Comment by ajosh0504** [comment]
We finally went with the user configuring only the parameter `anonymization` in the config file as faker/hash. For the hash based anonymization, the service makes a call to the Elasticsearch License API. For Elastic Cloud users, the `issued_to` in the return response is "Elastic Cloud". For Cloud users, the customer ID is then obtained from the Deployments API. For on-prem users, `issued-to` corresponds to the customer name itself. This is then used directly as the salt for hashing. This implementation has now been merged. Closing this.
